app/main.py

```diff
@@ -1,2 +1,224 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from fractions import Fraction
+from typing import Hashable, Any, Iterator, Optional, Iterable
+
+
 class Dictionary:
-    pass
+    @dataclass
+    class Node:
+        key: Hashable
+        value: Any
+        key_hash: int
+
+    INITIAL_CAPACITY = 8
+    RESIZE_THRESHOLD = Fraction(2, 3)
+    CAPACITY_MULTIPLIER = 2
+
+    def __init__(self) -> None:
+        self.capacity = self.INITIAL_CAPACITY
+        self.size = 0
+        self.hash_table: list[Dictionary.Node | None] = [None] * self.capacity
+
+    def _calculate_index(self, key: Hashable) -> int:
+        _key_hash = hash(key)
+        index = _key_hash % self.capacity
+
+        while (
+            self.hash_table[index] is not None
+            and (
+                self.hash_table[index].key != key
+                or self.hash_table[index].key_hash != _key_hash
+            )
+        ):
+            index = (index + 1) % self.capacity
+
+        return index
+
+    @property
+    def current_max_size(self) -> int:
+        return int(self.capacity * self.RESIZE_THRESHOLD)
+
+    def _resize_up(self) -> None:
+        old_hash_table = self.hash_table
+
+        self.capacity *= self.CAPACITY_MULTIPLIER
+        self.hash_table = [None] * self.capacity
+        self.size = 0
+
+        for node in filter(None, old_hash_table):
+            self.__setitem__(node.key, node.value)
+
+    def _resize_down(self) -> None:
+        old_hash_table = self.hash_table
+
+        self.capacity //= self.CAPACITY_MULTIPLIER
+        self.hash_table = [None] * self.capacity
+        self.size = 0
+
+        for node in filter(None, old_hash_table):
+            self.__setitem__(node.key, node.value)
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        if self.size + 1 > self.current_max_size:
+            self._resize_up()
+
+        index = self._calculate_index(key)
+        key_hash = hash(key)
+
+        if (
+            self.hash_table[index] is not None
+            and self.hash_table[index].key == key
+        ):
+            self.hash_table[index].value = value
+            return
+
+        if self.hash_table[index] is None:
+            self.size += 1
+
+        self.hash_table[index] = Dictionary.Node(key, value, key_hash)
+
+    def __getitem__(self, key: Hashable) -> Any:
+        index = self._calculate_index(key)
+
+        if self.hash_table[index] is None:
+            raise KeyError(f"Cannot find value for key: {key}")
+
+        return self.hash_table[index].value
+
+    def __delitem__(self, key: Hashable) -> None:
+        if self.size - 1 < self.current_max_size // self.CAPACITY_MULTIPLIER:
+            self._resize_down()
```

Comment:
Good job on implementing size down. But actually it would make sense to call it after an element has been removed